baseservice/base_service.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import signal
+import threading
+from abc import ABCMeta, abstractmethod
+from enum import Enum, unique
+from typing import Optional
+
+from baseservice.utils import Event
+
+
+@unique
+class ServiceState(Enum):
+    """
+    the states of the service: STARTING->(prepare_service)->STARTED->(run_service)->STOPPING->(finalize_service)->STOPPED
+    """
+    STARTING = 'STARTING'
+    STARTED = 'STARTED'
+    STOPPING = 'STOPPING'
+    STOPPED = 'STOPPED'
+
+
+class BaseService(metaclass=ABCMeta):
+    """
+    this class is the base class for all services
+    """
+
+    def __init__(self, *,
+                 name: str = None,
+                 should_stop_on_signal=True):
+        """
+
+        :param name: the name of this service. if None, the name of the type will be used
+        :param should_stop_on_signal: if True, the service will try to register SIGTERM and SIGINT on stop method.
+        """
+        self._should_stop_on_signal: bool = should_stop_on_signal
+        self._cancellation_token: threading.Event = threading.Event()
+        self._cancellation_token.set()  # service starts as not_running
+        self._logger = logging.getLogger(__name__)
+        self._state_changed_event: Event[ServiceState] = Event()
+        self._service_state = ServiceState.STOPPED
```

Comment:
Suggestion:
Shouldn't we have an 'initialized' service state for the first state of the service? 
To avoid confusions between service that haven't started even once, and other that have passed the starting state.

Reply:
I don't really see the difference.
whats the use case that justifies another state?